### PR TITLE
azure: remove disk locks per vm during attach/detach

### PR DESCRIPTION
**Pull request by aramase**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
- Changes the per VM lock logic for attach/detach operations.

Currently the per VM lock logic uses number of CPUs by default to lock which throttles number of concurrent attach/detach requests. With this updated locking logic, disk attach/detach for different nodes will be run concurrently. Disk attach/detach operations for the same node will be done sequentially.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
azure: update disk lock logic per vm during attach/detach to allow concurrent updates for different nodes.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
NONE
```

Test result -

```
Testing on a cluster with 1VMSS - 100 nodes

With the following changes -
Time taken for 10 disks (5 pods on same node + 5 pods on different nodes) - 9min
Time taken to scale from 10 disks to 60 disks - 6m36s
Time taken to detach and delete all 60 disks - 14m28s

Before the changes -
Time taken for 10 disks (5 pods on same node + 5 pods on different nodes) - 16min
Time taken to scale from 10 disks to 60 disks - 49m58s
```

### CODE REVIEW

**Comment by andyzhangx** [comment]
@aramase disk attach/detach lock cannot be removed now, is there a way to replace it with another better lock? This lock `uses number of CPUs by default`, we may change the default max thread num size.

**Comment by aramase** [comment]
@andyzhangx I'm going to test the behavior of removing the lock. If there are any issues, then yes we can investigate using the existing lock with different config or a different locking implementation.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_controller_common.go
Comment (andyzhangx) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_controller_common.go]: @aramase have you tried give a big number for NewHashed func? that could be simpler fix.

Comment (andyzhangx) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_controller_common.go]: I was wrong, your fix is based on string entry, it's a right fix.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_utils.go
Comment (andyzhangx) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_utils.go]: would you write a ut for this class?

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_utils.go
Comment (andyzhangx) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_utils.go]: Pls always use lowercase, otherwise if nodeName changed case, this lock may not work.
```
entry = strings.ToLower(entry)
```

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_utils.go
Comment (aramase) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_utils.go]: Have added unit tests for the class.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_utils.go
Comment (aramase) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_utils.go]: the entry should be decided by the caller. So added `strings.ToLower(string(nodeName))` at the caller level.

**Comment by aramase** [comment]
Test result -

```
Testing on a cluster with 1VMSS - 100 nodes

With the following changes -
Time taken for 10 disks (5 pods on same node + 5 pods on different nodes) - 9min
Time taken to scale from 10 disks to 60 disks - 6m36s
Time taken to detach and delete all 60 disks - 14m28s

Before the changes -
Time taken for 10 disks (5 pods on same node + 5 pods on different nodes) - 16min
Time taken to scale from 10 disks to 60 disks - 49m58s
```

The tests are done on a cluster with 2 vcpus. This means the length of mutexes is equal to the CPU num (2). In case of AKS, the master has 8vcpus from what @andyzhangx told. This would mean with the old keymutex, there could be 8 concurrent updates (attach/detach) in parallel provided there are no hash collisions. With the changes in this PR, all vm updates happen in parallel which reduces disk attach/detach time for numerous disks.

@andyzhangx PTAL!

/hold 
(adding a hold because want to get a approval from @khenidak too)

**Review by andyzhangx — APPROVED** [review]
/lgtm
/approve

**Comment by andyzhangx** [comment]
/hold cancel
since it's quite tight

**Comment by andyzhangx** [comment]
@aramase could you also cherry-pick this PR to old releases? thanks.

**Comment by aramase** [comment]
@andyzhangx will do!

**Pull request merged**
